File: ts/package/src/auth/handlers.ts
import axios from "axios";
import { SessionPost, type Session, type Sessioned } from "../utils/session.js";

export function scope() {
  return "/xnode-auth";
}

export type login_input = {
  baseUrl: string;
  user: string;
} & (
  | {
      signature: string;
      timestamp: string;
    }
  | {}
);
export type login_output = Session;
export async function login({
  baseUrl,
  ...loginParams
}: login_input): Promise<login_output> {
  const axiosInstance = axios.create({
    withCredentials: true, // Store cookies
  });

  await axiosInstance.post(`${baseUrl}${scope()}/api/login`, loginParams);

  return { axiosInstance, baseUrl };
}

export type logout_input = Sessioned<{}>;
export type logout_output = void;
export async function logout(input: logout_input): Promise<logout_output> {
  return SessionPost(input, scope(), "/api/logout");
}

File: rust/package/src/auth/handlers.rs
use reqwest::Client;
use serde::{Deserialize, Serialize};

use crate::utils::{Empty, Error, Session, SessionPostInput, SessionPostOutput, session_post};

pub fn scope() -> String {
    "/xnode-auth".to_string()
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct User {
    user: String,
    signature: Option<String>,
    timestamp: Option<String>,
}
impl User {
    pub fn new(user: String) -> Self {
        Self {
            user,
            signature: None,
            timestamp: None,
        }
    }

    pub fn with_signature(user: String, signature: String, timestamp: String) -> Self {
        Self {
            user,
            signature: Some(signature),
            timestamp: Some(timestamp),
        }
    }
}
#[derive(Debug, Clone, PartialEq)]
pub struct LoginInput {
    pub base_url: String,
    pub user: User,
}
pub type LoginOutput = Result<Session, Error>;
pub async fn login(input: LoginInput) -> LoginOutput {
    let client = Client::builder()
        .cookie_store(true)
        .build()
        .map_err(Error::ReqwestError)?;

    client
        .post(format!("{}{}/api/login", input.base_url, scope()))
        .json(&input.user)
        .send()
        .await
        .map_err(Error::ReqwestError)?;

    Ok(Session {
        reqwest_client: client,
        base_url: input.base_url,
    })
}

pub type LogoutInput<'a> = SessionPostInput<'a, Empty, Empty>;
pub type LogoutOutput = Empty;
pub async fn logout(input: LogoutInput<'_>) -> SessionPostOutput<LogoutOutput> {
    session_post(input, scope(), |_path| "/api/logout").await
}
